code/espurna/ota_asynctcp.ino

```diff
@@ -0,0 +1,220 @@
+/*
+
+ASYNC CLIENT OTA MODULE
+
+Copyright (C) 2016-2019 by Xose Pérez <xose dot perez at gmail dot com>
+
+*/
+
+#if OTA_CLIENT == OTA_CLIENT_ASYNCTCP
+
+// -----------------------------------------------------------------------------
+// Terminal OTA command
+// -----------------------------------------------------------------------------
+
+#if TERMINAL_SUPPORT || OTA_MQTT_SUPPORT
+
+#include <ESPAsyncTCP.h>
+#include "libs/URL.h"
+
+std::unique_ptr<AsyncClient> _ota_client = nullptr;
+unsigned long _ota_size = 0;
+bool _ota_connected = false;
+std::unique_ptr<URL> _ota_url = nullptr;
+
+const char OTA_REQUEST_TEMPLATE[] PROGMEM =
+    "GET %s HTTP/1.1\r\n"
+    "Host: %s\r\n"
+    "User-Agent: ESPurna\r\n"
+    "Connection: close\r\n"
+    "Content-Type: application/x-www-form-urlencoded\r\n"
+    "Content-Length: 0\r\n\r\n\r\n";
+
+void _otaClientOnDisconnect(void *s, AsyncClient *c) {
+
+    DEBUG_MSG_P(PSTR("\n"));
+
+    if (Update.end(true)){
+        DEBUG_MSG_P(PSTR("[OTA] Success: %u bytes\n"), _ota_size);
+        deferredReset(100, CUSTOM_RESET_OTA);
+    } else {
+        #ifdef DEBUG_PORT
+            Update.printError(DEBUG_PORT);
+        #endif
+        eepromRotate(true);
+    }
+
+    DEBUG_MSG_P(PSTR("[OTA] Disconnected\n"));
+
+    _ota_connected = false;
+    _ota_url = nullptr;
+    _ota_client = nullptr;
+
+}
+
+void _otaClientOnTimeout(void *s, AsyncClient *c, uint32_t time) {
+    _ota_connected = false;
+    _ota_url = nullptr;
+    _ota_client->close(true);
+}
+
+void _otaClientOnData(void * arg, AsyncClient * c, void * data, size_t len) {
+
+    char * p = (char *) data;
+
+    if (_ota_size == 0) {
+
+        Update.runAsync(true);
+        if (!Update.begin((ESP.getFreeSketchSpace() - 0x1000) & 0xFFFFF000)) {
+            #ifdef DEBUG_PORT
+                Update.printError(DEBUG_PORT);
+            #endif
+        }
+
+        p = strstr((char *)data, "\r\n\r\n") + 4;
+        len = len - (p - (char *) data);
+
+    }
+
+    if (!Update.hasError()) {
+        if (Update.write((uint8_t *) p, len) != len) {
+            #ifdef DEBUG_PORT
+                Update.printError(DEBUG_PORT);
+            #endif
+        }
+    }
+
+    _ota_size += len;
+    DEBUG_MSG_P(PSTR("[OTA] Progress: %u bytes\r"), _ota_size);
+
+    delay(0);
+
+}
+
+void _otaClientOnConnect(void *arg, AsyncClient *client) {
+
+    #if ASYNC_TCP_SSL_ENABLED
+        if (443 == _ota_url->port) {
+            uint8_t fp[20] = {0};
+            sslFingerPrintArray(getSetting("otafp", OTA_FINGERPRINT).c_str(), fp);
```

Comment:
Should have a check for ``otaScCheck`` here as well, and skip the fingerprinting if it's ``SECURE_CLIENT_CHECK_NONE``.